Feat/add rune impr

valeryV99 opened the pull request:
https://www.notion.so/254362757c98814193bdf2683a2f8535?v=254362757c98811fb5a9000c47b44575&p=25d362757c9880efbfbced125a7a6e60&pm=s

Refactored, add rune component

valeryV99 committed:
feat: refactor AddRuneDialog to enhance ERC20 handling and error reporting

- Replaced `useEdictRune` with `useAddRuneERC20` for ERC20 integration.
- Implemented new `useERC20Rune` hook to monitor ERC20 readiness.
- Improved transaction state handling with `hasBroadcast` and `isErc20Ready` checks.
- Enhanced error feedback with dynamic error message display in UI.
- Simplified fee transfer logic in `onConfirm` function.

valeryV99 committed:
Merge remote-tracking branch 'origin/staging' into feat/add-rune-impr

valeryV99 committed:
feat: improve AddRuneDialog with enhanced polling and error handling

- Added retry mechanism and state tracking for ERC20 readiness polling.
- Enhanced error handling with dynamic error message updates in the UI.
- Fixed refetch intervals and retries for `useERC20Rune` when broadcasting transactions.
- Refactored transaction state logic for better readability and reliability.

valeryV99 committed:
feat: update AddRuneDialog ERC20 polling logic and add tests

- Renamed `shouldPollEvm` to `shouldPollErc` for clarity.
- Updated all references to align with the new naming convention.
- Added comprehensive unit tests for `AddRuneDialog` to cover various states:
  - Initial rendering.
  - Transaction confirmation and explorer link display.
  - Success and error states during ERC20 polling.
  - Handling retries and fallback errors for zero-address cases.

Georgyskr commented on apps/web/src/features/runes/ui/add-rune-dialog/AddRuneDialog.tsx:
```diff
@@ -53,39 +58,73 @@ export const AddRuneDialog = ({ onClose, ...rest }: AddRuneDialogProps) => {
+    error?.message ||
+    (retriesExhaustedNoAddress ? 'Rune addition error' : null);
+
   const { data: edictFee } = useQuery({
```
I don't think we need to query edict fee at all anymore?

valeryV99 replied:
discussed

Georgyskr commented on apps/web/src/features/runes/ui/add-rune-dialog/AddRuneDialog.tsx:
```diff
@@ -53,39 +58,73 @@ export const AddRuneDialog = ({ onClose, ...rest }: AddRuneDialogProps) => {
+  const errorMessage =
+    erc20Error?.message ||
+    error?.message ||
+    (retriesExhaustedNoAddress ? 'Rune addition error' : null);
```
Let's have the following wording
Failed to add the token. Please try doing it once again.

valeryV99 replied:
done

Georgyskr commented on apps/web/src/features/runes/ui/add-rune-dialog/AddRuneDialog.tsx:
```diff
@@ -53,39 +58,73 @@ export const AddRuneDialog = ({ onClose, ...rest }: AddRuneDialogProps) => {
   } = useWaitForTransaction();
   const { network } = useConfig();
 
   const feeRate = useBTCFeeRate();
```
I don't think we need this anymore

valeryV99 replied:
discussed

Georgyskr commented:
@opcode0x need some help with tests review while I'm still getting a hang of things

valeryV99 committed:
fix: improve AddRuneDialog error message and update tests

- Updated error message for retriesExhaustedNoAddress to provide clearer guidance.
- Adjusted corresponding test case to validate the updated error message.

opcode0x commented on apps/web/src/features/runes/ui/add-rune-dialog/AddRuneDialog.tsx:
```diff
@@ -53,39 +58,75 @@ export const AddRuneDialog = ({ onClose, ...rest }: AddRuneDialogProps) => {
+  const isErc20Ready = !!erc20Address && erc20Address !== zeroAddress;
+
+  // Track refetch attempts by observing dataUpdatedAt; after 6 updates with zero address -> error
+  const [refetchCount, setRefetchCount] = useState(0);
```
remove refetchCount from react state

valeryV99 committed:
refactor(AddRuneDialog): replace useState with useRef for refetchCount

- Updated `refetchCount` to use `useRef` instead of `useState` for consistent and mutable reference across renders.
- Adjusted logic in `useEffect` and associated conditions to accommodate the `useRef` change.
